01-overview-of-ml-systems.md

```diff
@@ -0,0 +1,110 @@
+
+
+**2. Complex patterns: there are patterns to learn, and they are complex.**
+
+- If you have data but there is no pattern to learn, then an ML approach doesn't make sense. For example, it doesn't make sense to build an ML system to predict the outcome for a fair die. 
+
+- If there are patterns but they are simple, it probably makes more sense to use a pure software engineering solution for the problem (e.g. don't build an ML system to look for zip codes based on address, just use a lookup table).
+
+  
+**3. The problem can be framed as a prediction problem.**
+
+- Supervised ML models make predictions. If the whole problem  (or part of it) cannot be framed as a prediction problem, supervised ML is not the right tool.
+- Many computationally intensive problems are getting reframed as prediction problems. An example of this is light rendering in video games. Instead of doing the full expensive computation of what the lighting should be exactly, a regression ML model can be used to get a "good enough" approximation. In this case training the model is cheaper than doing the calculation.
+
+
+4. **Unseen data shares patterns with the train data.**
+
+- Your model only works well if the patterns in the production data are the same than the ones in the training data.
+- *If the data is unseen, how do we know what distribution it comes from?* We don't, we just make a reasonable assumption that it is. We can retrospectively check if our assumption was true by monitoring in production (chapter 8) and testing in production (chapter 9).
+
+
+
+## Problem characteristics that will make ML solutions especially useful
+These are not necessary characteristics, but having one or multiple of them make it much more likely that an ML approach will succeed for a problem. 
+
+1. **It is a repetitive task:**  if a task is repetitive, patterns will be repeated multiple times and the ML models will be able to learn better.
+2. **The cost of wrong predictions is cheap**: ML models will make wrong predictions. When mistakes are cheap, stakes are lower, making it easier to build and deploy ML models. Models can still be built when the consequences of a wrong prediction are catastrophic if on average the benefits of correct predictions outweigh the costs of the wrong ones. Self-driving cars are a good example for this when considering that automated cars on average are safer than cars driven by humans.
+3. **The task happens at scale:** deploying ML models to production is hard. This work is more justifiable when the model will be used to make lots of predictions. Also, if the task happens at scale, there will be lots of data to train from.
+4. **The patterns are constantly changing:** ML shines against traditional software rules when patterns change frequently, causing hardcoded rules get outdated too fast.  Changes in spam trends and culture trends are examples of changing patterns.
```

Comment:
My work in Amazon fulfillment falls into this category. We train a model every two weeks for each warehouse since the defect patterns are subject to personnel and seasonality.